File: Documentation/6.0/Samples/python/ClientApi/Session/Querying/HowToProjectQueryResults.py
import datetime
from typing import List

from ravendb import QueryData
from ravendb.infrastructure.orders import Company, Address, Order, OrderLine

from examples_base import ExampleBase, Employee


class HowToProjectQueryResults(ExampleBase):
    def setUp(self):
        super().setUp()
        with self.embedded_server.get_document_store("ProjectQueryResults") as store:
            with store.open_session() as session:
                self.add_orders(session)
                self.add_companies(session)
                self.add_employees(session)

    def test_examples(self):
        with self.embedded_server.get_document_store("ProjectQueryResults") as store:
            with store.open_session() as session:
                # region projections_1

                class CompanyNameCityAndCountry:
                    def __init__(self, name: str = None, city: str = None, country: str = None):
                        self.name = name
                        self.city = city
                        self.country = country

                query_data = QueryData(["name", "address.city", "address.country"], ["name", "city", "country"])
                results = list(
                    session.query(object_type=Company).select_fields_query_data(CompanyNameCityAndCountry, query_data)
                )

                # Each resulting object in the list is not a 'Company' entity, it is a new object containing ONLY the
                # fields specified in the query_data
                # endregion

            with store.open_session() as session:
                # region projections_2
                class OrderShippingAddressAndProductNames:
                    def __init__(self, ship_to: str = None, product_names: List[str] = None):
                        self.ship_to = ship_to
                        self.product_names = product_names

                # Retrieve all product names from the Lines array in an Order document
                query_data = QueryData(["ship_to", "lines[].product_name"], ["ship_to", "product_names"])

                projected_results = list(
                    session.query(object_type=Order).select_fields_query_data(
                        OrderShippingAddressAndProductNames, query_data
                    )
                )
                # endregion

            with store.open_session() as session:
                # region projections_3
                class EmployeeFullName:
                    def __init__(self, full_name: str):
                        self.full_name = full_name

                # Use custom function in query data or raw query
                query_data = QueryData.custom_function("o", "{ full_name: o.first_name + ' ' + o.last_name }")
                projected_results = list(
                    session.query(object_type=Employee).select_fields_query_data(EmployeeFullName, query_data)
                )
                # endregion

            with store.open_session() as session:
                # region projections_4
                class ProductsRaport:
                    def __init__(
                        self, total_products: int = None, total_discounted_products: int = None, total_price: int = None
                    ):
                        self.total_products = total_products
                        self.total_discounted_products = total_discounted_products
                        self.total_price = total_price

                # Use custom function in query data or raw query
                query_data = QueryData.custom_function(
                    "o",
                    "{"
                    "total_products: o.lines.length,"
                    " total_discounted_products: o.lines.filter((line) => line.discount > 0).length,"
                    " total_price: o.lines.reduce("
                    "(accumulator, line) => accumulator + line.price_per_unit * line.quantity, 0) "
                    "}",
                )
                projected_results = list(
                    session.query(object_type=Order).select_fields_query_data(ProductsRaport, query_data)
                )
                # endregion

            with store.open_session() as session:
                # region projections_7
                class EmployeeAgeDetails:
                    def __init__(self, day_of_birth: str = None, month_of_birth: str = None, age: str = None):
                        self.day_of_birth = day_of_birth
                        self.month_of_birth = month_of_birth
                        self.age = age

                # Use custom function in query data or raw query
                results = session.advanced.raw_query(
                    "from Employees as e select {"
                    ' "day_of_birth : new Date(Date.parse(e.birthday)).getDate(),'
                    " month_of_birth : new Date(Date.parse(e.birthday)).getMonth() + 1,"
                    " age : new Date().getFullYear() - new Date(Date.parse(e.birthday)).getFullYear()"
                    "}",
                    EmployeeAgeDetails,
                )
                # endregion

            with store.open_session() as session:
                # region projections_8
                class EmployeeBirthdayAndName:
                    def __init__(self, date: str = None, name: str = None):
                        self.date = date
                        self.name = name

                # Use custom function in query data or raw query
                results = list(
                    session.advanced.raw_query(
                        "from Employees as e select {"
                        "date: new Date(Date.parse(e.birthday)),"
                        "name: e.first_name.substr(0,3)"
                        "}",
                        EmployeeBirthdayAndName,
                    )
                )
                # endregion

            with store.open_session() as session:

                class EmployeeNameAndMetadata:
                    def __init__(self, name: str = None, metadata: str = None):
                        self.name = name
                        self.metadata = metadata

                # region projections_9
                projected_results = list(
                    session.advanced.raw_query(
                        "from Employees as e "
                        + "select {"
                        + "     name : e.first_name, "
                        + "     metadata : getMetadata(e)"
                        + "}",
                        EmployeeNameAndMetadata,
                    )
                )
                # endregion

            with store.open_session() as session:
                # region projections_10
                projected_results = list(
                    session.query(object_type=Company)
                    # Pass the projection class
                    .select_fields(ContactDetails)
                )
                # endregion

            with store.open_session() as session:
                # region projections_12
                # Lets define an array with the field names that will be projected
                # (its optional, you can pass field names as args loosely)
                projection_fields = ["name", "phone"]
                # Make a query
                projected_results = list(
                    session.advanced.document_query(object_type=Company)
                    # Call 'select_fields'
                    # Pass the projection class type & the fields to be projected from it
                    .select_fields(ContactDetails, *projection_fields)
                )

                # Each resulting object in the list is not a 'Company' entity
                # it is an object of type 'ContactDetails' containing data ONLY for the specified fields
                # endregion

            with store.open_session() as session:
                # region projections_13
                # For example:
                query_data = QueryData(["name"], ["funny_name"])
                try:
                    projected_results = list(
                        session.query(object_type=Company)
                        # Make a first projection
                        .select_fields(ContactDetails)
                        # A second projection is not supported and will raise an error
                        .select_fields_query_data(CompanyNameCityAndCountry, query_data)
                    )
                except Exception as e:
                    pass
                    # The following exception will be raised:
                    # "Projection is already done. You should not project your result twice."
                # endregion


# region projection_class
class ContactDetails:
    # The projection class contains field names from the 'Company' document
    def __init__(self, name: str = None, phone: str = None, fax: str = None):
        self.name = name
        self.phone = phone
        self.fax = fax


# endregion

File: Documentation/6.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Querying/HowToProjectQueryResults.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Linq;
using Raven.Client.Documents.Queries;
using Raven.Documentation.Samples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Querying
{
    public class ProjectQueryResults
    {
        public async Task Examples()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region projections_1
                    var projectedResults = session
                         // Make a dynamic query on the Companies collection
                        .Query<Company>()
                         // Call Select to define the new structure that will be returned per Company document
                        .Select(x => new
                        {
                            Name = x.Name,
                            City = x.Address.City,
                            Country = x.Address.Country
                        })
                        .ToList();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is a new object containing ONLY the fields specified in the Select.
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_1_async
                    var projectedResults = await asyncSession
                        // Make a dynamic query on the Companies collection
                        .Query<Company>()
                        // Call Select to define the new structure that will be returned per Company document
                        .Select(x => new {Name = x.Name, City = x.Address.City, Country = x.Address.Country})
                        .ToListAsync();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is a new object containing ONLY the fields specified in the Select.
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_2
                    var projectedResults = session
                        .Query<Order>()
                        .Select(x => new
                        {
                            ShipTo = x.ShipTo,
                            // Retrieve all product names from the Lines array in an Order document
                            ProductNames = x.Lines.Select(y => y.ProductName)
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_2_async
                    var projectedResults = await asyncSession
                        .Query<Order>()
                        .Select(x => new
                        {
                            ShipTo = x.ShipTo,
                            // Retrieve all product names from the Lines array in an Order document
                            ProductNames = x.Lines.Select(y => y.ProductName)
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_3
                    var projectedResults = session
                        .Query<Employee>()
                        .Select(x => new
                        {
                            // Any expression can be provided for the projected content
                            FullName = x.FirstName + " " + x.LastName
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_3_async
                    var projectedResults = await asyncSession
                        .Query<Employee>()
                        .Select(x => new
                        {
                            // Any expression can be provided for the projected content
                            FullName = x.FirstName + " " + x.LastName
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_4
                    var projectedResults = session
                        .Query<Order>()
                        .Select(x => new
                        {
                            // Any calculations can be done within a projection
                            TotalProducts = x.Lines.Count,
                            TotalDiscountedProducts = x.Lines.Count(x => x.Discount > 0),
                            TotalPrice = x.Lines.Sum(l => l.PricePerUnit * l.Quantity)
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_4_async
                    var projectedResults = await asyncSession
                        .Query<Order>()
                        .Select(x => new
                        {
                            // Any calculations can be done within a projection
                            TotalProducts = x.Lines.Count,
                            TotalDiscountedProducts = x.Lines.Count(x => x.Discount > 0),
                            TotalPrice = x.Lines.Sum(l => l.PricePerUnit * l.Quantity)
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_5
                    // Use LINQ query syntax notation
                    var projectedResults = (from e in session.Query<Employee>()
                        // Define a function
                        let format = (Func<Employee, string>)(p => p.FirstName + " " + p.LastName)
                        select new
                        {
                            // Call the function from the projection
                            FullName = format(e)
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_5_async
                    // Use LINQ query syntax notation
                    var projectedResults = await (from e in asyncSession.Query<Employee>()
                        // Define a function
                        let format = (Func<Employee, string>)(p => p.FirstName + " " + p.LastName)
                        select new
                        {
                            // Call the function from the projection
                            FullName = format(e)
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_6
                    // Use LINQ query syntax notation
                    var projectedResults = (from o in session.Query<Order>()
                        // Use RavenQuery.Load to load the related Company document
                        let c = RavenQuery.Load<Company>(o.Company)
                        select new
                        {
                            CompanyName = c.Name, // info from the related Company document
                            ShippedAt = o.ShippedAt // info from the Order document
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_6_async
                    // Use LINQ query syntax notation
                    var projectedResults = (from o in asyncSession.Query<Order>()
                        // Use RavenQuery.Load to load the related Company document
                        let c = RavenQuery.Load<Company>(o.Company)
                        select new
                        {
                            CompanyName = c.Name, // info from the related Company document
                            ShippedAt = o.ShippedAt // info from the Order document
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_7
                    var projectedResults = session
                        .Query<Employee>()
                        .Select(e => new
                        {
                            DayOfBirth = e.Birthday.Day,
                            MonthOfBirth = e.Birthday.Month,
                            Age = DateTime.Today.Year - e.Birthday.Year
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_7_async
                    var projectedResults = await asyncSession
                        .Query<Employee>()
                        .Select(e => new
                        {
                            DayOfBirth = e.Birthday.Day,
                            MonthOfBirth = e.Birthday.Month,
                            Age = DateTime.Today.Year - e.Birthday.Year
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_8
                    var projectedResults = session.Query<Employee>()
                        .Select(e => new
                        {
                            // Provide a JavaScript expression to the RavenQuery.Raw method
                            Date = RavenQuery.Raw<DateTime>("new Date(Date.parse(e.Birthday))"),
                            Name = RavenQuery.Raw(e.FirstName, "substr(0, 3)")
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_8_async
                    var projectedResults = await asyncSession.Query<Employee>()
                        .Select(e => new
                        {
                            // Provide a JavaScript expression to the RavenQuery.Raw method
                            Date = RavenQuery.Raw<DateTime>("new Date(Date.parse(e.Birthday))"),
                            Name = RavenQuery.Raw(e.FirstName, "substr(0, 3)")
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_9
                    var projectedResults = session.Query<Employee>()
                        .Select(e => new
                        {
                            Name = e.FirstName, Metadata = RavenQuery.Metadata(e) // Get the metadata
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_9_async
                    var projectedResults = await asyncSession.Query<Employee>()
                        .Select(e => new
                        {
                            Name = e.FirstName, Metadata = RavenQuery.Metadata(e) // Get the metadata
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_10
                    var projectedResults = session
                        .Query<Company>()
                        // Call 'ProjectInto' instead of using 'Select'
                        // Pass the projection class
                        .ProjectInto<ContactDetails>()
                        .ToList();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is an object of type 'ContactDetails'.
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_10_async
                    var projectedResults = await asyncSession
                        .Query<Company>()
                        // Call 'ProjectInto' instead of using 'Select'
                        // Pass the projection class
                        .ProjectInto<ContactDetails>()
                        .ToListAsync();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is an object of type 'ContactDetails'.
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_11
                    // Make a dynamic DocumentQuery
                    var projectedResults = session.Advanced
                        .DocumentQuery<Company>()
                        // Call 'SelectFields'
                        // Pass the projection class type
                        .SelectFields<ContactDetails>()
                        .ToList();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is an object of type 'ContactDetails'.
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_11_async
                    // Make a dynamic DocumentQuery
                    var projectedResults = await asyncSession.Advanced
                        .AsyncDocumentQuery<Company>()
                        // Call 'SelectFields'
                        // Pass the projection class type
                        .SelectFields<ContactDetails>()
                        .ToListAsync();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is an object of type 'ContactDetails'.
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_12
                    // Define an array with the field names that will be projected
                    var projectionFields = new string[]
                    {
                        // Fields from 'ContactDetails' class:
                        "Name", "Phone"
                    };

                    // Make a dynamic DocumentQuery
                    var projectedResults = session.Advanced
                        .DocumentQuery<Company>()
                        // Call 'SelectFields'
                        // Pass the projection class type & the fields to be projected from it
                        .SelectFields<ContactDetails>(projectionFields)
                        .ToList();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is an object of type 'ContactDetails' containing data ONLY for the specified fields.
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_12_async
                    // Define an array with the field names that will be projected
                    var projectionFields = new string[]
                    {
                        // Fields from 'ContactDetails' class:
                        "Name", "Phone"
                    };

                    // Make a dynamic DocumentQuery
                    var projectedResults = await asyncSession.Advanced
                        .AsyncDocumentQuery<Company>()
                        // Call 'SelectFields'
                        // Pass the projection class type & the fields to be projected from it
                        .SelectFields<ContactDetails>(projectionFields)
                        .ToListAsync();

                    // Each resulting object in the list is Not a 'Company' entity,
                    // it is an object of type 'ContactDetails' containing data ONLY for the specified fields.
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_13
                    // For example:
                    try
                    {
                        var projectedResults = session
                            .Query<Company>()
                            // Make first projection
                            .ProjectInto<ContactDetails>()
                            // A second projection is not supported and will throw
                            .Select(x => new {Name = x.Name})
                            .ToList();
                    }
                    catch (Exception e)
                    {
                        // The following exception will be thrown:
                        // "Projection is already done. You should not project your result twice."
                    }
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region projections_14
                    var projectedResults = session
                        .Query<Product>()
                         // NOTE:
                         // While the following 'Include' line compiles,
                         // the related Supplier document will NOT BE INCLUDED in the query results,
                         // because 'Supplier' is not one of the projected fields in the 'Select' clause.
                        .Include(x => x.Supplier)
                        .Select(x => new
                        {
                            Name = x.Name,
                            ProductCategory = x.Category
                        })
                         // The related Category document WILL BE INCLUDED in the query results,
                         // since 'ProductCategory' is one of the projected fields.
                        .Include(x => x.ProductCategory)
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_14_async
                    var projectedResults = await asyncSession
                        .Query<Product>()
                         // NOTE:
                         // While the following 'Include' line compiles,
                         // the related Supplier document will NOT BE INCLUDED in the query results,
                         // because 'Supplier' is not one of the projected fields in the 'Select' clause.
                        .Include(x => x.Supplier)
                        .Select(x => new
                        {
                            Name = x.Name,
                            ProductCategory = x.Category
                        })
                         // The related Category document WILL BE INCLUDED in the query results,
                         // since 'ProductCategory' is one of the projected fields.
                        .Include(x =>x.ProductCategory)
                        .ToListAsync();
                    #endregion
                }
            }
        }

        #region projections_class
        public class ContactDetails
        {
            // The projection class contains field names from the 'Company' document
            public string Name { get; set; }
            public string Phone { get; set; }
            public string Fax { get; set; }
        }
        #endregion
    }
}
